### ISSUE TITLE
GROBID service crashes with more than one thread

### ISSUE THREAD

**Issue by FraDePace** [kermitt2/grobid#436]
Hi,

I tried to parse 136 papers using the Grobid Service and the Java GROBID client (https://github.com/kermitt2/grobid-client-java).

Unfortunately, every time I try to process the papers using more than one thread, the server crashes.
This behaviour is independent from the type of GROBID service (processFulltextDocument, processHeaderDocument, processReferences). Moreover, if I use only one thread, the server parses the paper without incurring in any problems.

This is my procedure:
1. start the server: gradlew run
2. start the client: java -jar target/org.grobid.client-0.5.4-SNAPSHOT.one-jar.jar -in path-to-INOUT-pdf -out path-to-OUTPUT-pdf -exe GROBID-SERVICE -n numberOfThread

I am using GROBID 0.5.5 on Windows 10.
If numberOfThread is greater than 1, the server starts to parse the papers but suddenly the following error appears (I'm using  java version "1.8.0_211"):

> Task :grobid-service:run FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':grobid-service:run'.
> Process 'command 'C:\Program Files\Java\jdk1.8.0_181\bin\java.exe'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.4.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 1m 38s
6 actionable tasks: 1 executed, 5 up-to-date

Any helps would be really appreciated.
Thanks in advance

**Comment by lfoppiano** [comment]
Dear @FraDePace, your problem might be windows related, not sure. Could you please past here the result of `./gradlew run --stacktrace --info`?

**Comment by kermitt2** [comment]
@FraDePace in complement you should have some server logs under `grobid/logs/grobid-service.log` which could be useful for us.

**Comment by FraDePace** [comment]
@lfoppiano this is the result of ./gradlew run --stacktrace --info
[GROBID_stacktrace.txt](https://github.com/kermitt2/grobid/files/3284594/GROBID_stacktrace.txt)

@kermitt2 this is the grobid-service.log
[grobid-service.log](https://github.com/kermitt2/grobid/files/3284591/grobid-service.log)

It seems that if the server fails to parse a paper, the entire process crashes.

**Comment by lfoppiano** [comment]
@FraDePace could you try to run the pdfalto command to any pdf, it's located in `grobid-home/pdf2xml/pdfalto`? 
Could you also, as a test, replace the content of `grobid-home/pdf2xml/win-64/` with the content of `grobid-home/pdf2xml/win-64/pdfalto`?

@kermitt2, looks like that pdfalto is in a subdirectory too much `grobid-home/pdf2xml/pdfalto`.. in DocumentSource I don't see anything specific there for Windows...

**Comment by kermitt2** [comment]
@lfoppiano this was a modification from Achraf because pdf2xml and pdfalto was depending on different versions of cygwin if I remember well - however I dont see under `grobid-home/pdf2xml/win-64/pdfalto` the other libraries for local linking, so I don't think this change can work.

**Comment by lfoppiano** [comment]
@kermitt2 I imagined, but but the path is wrong then... isn't it?

**Comment by kermitt2** [comment]
In `DocumentSource.java` line 78, we have a path change specific to Windows: 

```java
if (SystemUtils.IS_OS_WINDOWS) {
     //pdfalto executable are separated to avoid dll conflicts
     pdfToXml.append(File.separator +"pdfalto");
}
```

**Comment by lfoppiano** [comment]
OK, sorry. I was looking a the wrong branch... 🤦‍♂ 

@FraDePace ignore my second comment `Could you also, as a test, replace the content of grobid-home/pdf2xml/win-64/ with the content of grobid-home/pdf2xml/win-64/pdfalto?`.

**Comment by kermitt2** [comment]
One solution would be to remove entirely pdf2xml, but the problem is that it will make the current `grobid-home` not compatible with grobid jar libraries prior to 0.5.5 - other grobid modules running with older grobid lib. would fail with current grobid-home.

**Comment by kermitt2** [comment]
But the problem here is very likely due to the `pdfalto_server` under `grobid-home/pdf2xml/win-64/pdfalto`: with 1 thread `pdfalto` is used, with several threads it's `pdfalto_server`.

**Comment by FraDePace** [comment]
@lfoppiano this is the result of pdfalto.bat pdf
[pdfalto_bat.txt](https://github.com/kermitt2/grobid/files/3297704/pdfalto_bat.txt)

this is the result of pdfalto_server.bat pdf
[pdfalto_server_bat.txt](https://github.com/kermitt2/grobid/files/3297707/pdfalto_server_bat.txt)

**Comment by lfoppiano** [comment]
@FraDePace do you get output files from the commands? Could you attach them?

**Comment by FraDePace** [comment]
@lfoppiano result and output files of pdfalto.bat pdf
[pdfalto_bat.txt](https://github.com/kermitt2/grobid/files/3300230/pdfalto_bat.txt)
[output_pdfalto.zip](https://github.com/kermitt2/grobid/files/3300231/output_pdfalto.zip)

result and output files of pdfalto_server.bat pdf
[pdfalto_server_bat.txt](https://github.com/kermitt2/grobid/files/3300232/pdfalto_server_bat.txt)
[output_pdfalto_server.zip](https://github.com/kermitt2/grobid/files/3300234/output_pdfalto_server.zip)

**Comment by FraDePace** [comment]
UPDATE
not sure if it can be useful, but I face the same problem using grobid 0.5.4

**Comment by lfoppiano** [comment]
@FraDePace that's very useful, it means that PdfAlto is not the cause... We need to investigate more using windows... 

I suggest, perhaps, to consider the use of the docker container http://grobid.readthedocs.io/en/latest/Grobid-docker/ (remember to give at least 4Gb of RAM to the docker machine). Would be that possible?

**Comment by FraDePace** [comment]
@lfoppiano thanks, I'll try!

**Comment by lfoppiano** [comment]
@FraDePace did you manage to get it run? Any feedback?

**Comment by FraDePace** [comment]
@lfoppiano sorry for the late response. I did not try with a docker container but I got it run using the grobid service on a OSX Mojave :)

**Comment by kermitt2** [comment]
mmm given that everything is working fine with one thread anyway on Windows, it's normal that `pdfalto` and its output are okay when using the (non parallel) command line. It does not mean that the issue does not come from pdfalto when used in parallel mode. But what's nice if that we know it's not specific to `pdfalto_server`. In the output, we see that there are dependency issues with png for instance in the built version, but that's another issue.

We have this in the provided logs:

> ERROR [2019-06-13 07:03:22,731] org.grobid.service.process.GrobidRestProcessFiles: An unexpected exception occurs. 
! org.grobid.core.exceptions.GrobidException: [NO_BLOCKS] PDF parsing resulted in empty content

so it's still an indication about an issue with pdfalto or the way temporary files are managed when using multiple threads on Windows. Unfortunately I can't reproduce the problem because I don't have Windows machine, but I will try to find a way :/ 

It would be nice to rebuild pdfalto for windows with static dependencies if possible I think.
